// @remotion/cli/dist/editor/components/Timeline/TimelineInOutPointerHandle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineInOutPointerHandle = exports.outPointerHandle = exports.inPointerHandle = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const get_left_of_timeline_slider_1 = require("../../helpers/get-left-of-timeline-slider");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
const line = {
    height: '100%',
    width: 1,
    position: 'absolute',
    backgroundColor: 'rgba(255, 255, 255, 0.1)',
    cursor: 'ew-resize',
    paddingLeft: 1,
    paddingRight: 1,
};
exports.inPointerHandle = (0, react_1.createRef)();
exports.outPointerHandle = (0, react_1.createRef)();
const TimelineInOutPointerHandle = ({ dragging, type, atFrame }) => {
    const timelineWidth = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const videoConfig = (0, remotion_1.useVideoConfig)();
    if (timelineWidth === null) {
        throw new Error('Timeline width is null');
    }
    const style = (0, react_1.useMemo)(() => {
        return {
            ...line,
            backgroundColor: dragging
                ? 'rgba(255, 255, 255, 0.7)'
                : 'rgba(255, 255, 255, 0.1)',
            transform: `translateX(${(0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(atFrame, videoConfig.durationInFrames, timelineWidth)}px)`,
        };
    }, [atFrame, dragging, timelineWidth, videoConfig.durationInFrames]);
    return ((0, jsx_runtime_1.jsx)("div", { ref: type === 'in' ? exports.inPointerHandle : exports.outPointerHandle, style: style }));
};
exports.TimelineInOutPointerHandle = TimelineInOutPointerHandle;
