// @remotion/player/dist/cjs/event-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThumbnailEmitter = exports.PlayerEmitter = void 0;
class PlayerEmitter {
    constructor() {
        this.listeners = {
            ended: [],
            error: [],
            pause: [],
            play: [],
            ratechange: [],
            scalechange: [],
            seeked: [],
            timeupdate: [],
            frameupdate: [],
            fullscreenchange: [],
            volumechange: [],
            mutechange: [],
        };
    }
    addEventListener(name, callback) {
        this.listeners[name].push(callback);
    }
    removeEventListener(name, callback) {
        this.listeners[name] = this.listeners[name].filter((l) => l !== callback);
    }
    dispatchEvent(dispatchName, context) {
        this.listeners[dispatchName].forEach((callback) => {
            callback({ detail: context });
        });
    }
    dispatchSeek(frame) {
        this.dispatchEvent('seeked', {
            frame,
        });
    }
    dispatchVolumeChange(volume) {
        this.dispatchEvent('volumechange', {
            volume,
        });
    }
    dispatchPause() {
        this.dispatchEvent('pause', undefined);
    }
    dispatchPlay() {
        this.dispatchEvent('play', undefined);
    }
    dispatchEnded() {
        this.dispatchEvent('ended', undefined);
    }
    dispatchRateChange(playbackRate) {
        this.dispatchEvent('ratechange', {
            playbackRate,
        });
    }
    dispatchScaleChange(scale) {
        this.dispatchEvent('scalechange', {
            scale,
        });
    }
    dispatchError(error) {
        this.dispatchEvent('error', {
            error,
        });
    }
    dispatchTimeUpdate(event) {
        this.dispatchEvent('timeupdate', event);
    }
    dispatchFrameUpdate(event) {
        this.dispatchEvent('frameupdate', event);
    }
    dispatchFullscreenChange(event) {
        this.dispatchEvent('fullscreenchange', event);
    }
    dispatchMuteChange(event) {
        this.dispatchEvent('mutechange', event);
    }
}
exports.PlayerEmitter = PlayerEmitter;
class ThumbnailEmitter {
    constructor() {
        this.listeners = {
            error: [],
        };
    }
    addEventListener(name, callback) {
        this.listeners[name].push(callback);
    }
    removeEventListener(name, callback) {
        this.listeners[name] = this.listeners[name].filter((l) => l !== callback);
    }
    dispatchEvent(dispatchName, context) {
        this.listeners[dispatchName].forEach((callback) => {
            callback({ detail: context });
        });
    }
    dispatchError(error) {
        this.dispatchEvent('error', {
            error,
        });
    }
}
exports.ThumbnailEmitter = ThumbnailEmitter;
