// @remotion/cli/dist/editor/components/RenderQueue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderQueue = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const layout_1 = require("../layout");
const is_menu_item_1 = require("../Menu/is-menu-item");
const context_1 = require("./context");
const RenderQueueItem_1 = require("./RenderQueueItem");
const separatorStyle = {
    borderBottom: `1px solid ${colors_1.BORDER_COLOR}`,
};
const errorExplanation = {
    fontSize: 14,
    color: colors_1.LIGHT_TEXT,
    fontFamily: 'sans-serif',
    lineHeight: 1.5,
};
const explainer = {
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    padding: '0 12px',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
    background: colors_1.BACKGROUND,
};
const renderQueue = {
    background: colors_1.BACKGROUND,
    flex: 1,
    overflowY: 'auto',
};
const RenderQueue = () => {
    const connectionStatus = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx).type;
    const { jobs } = (0, react_1.useContext)(context_1.RenderQueueContext);
    const jobCount = jobs.length;
    if (connectionStatus === 'disconnected') {
        return ((0, jsx_runtime_1.jsxs)("div", { style: explainer, children: [(0, jsx_runtime_1.jsx)(layout_1.Spacing, { y: 5 }), (0, jsx_runtime_1.jsx)("div", { style: errorExplanation, children: "The studio server has disconnected." }), (0, jsx_runtime_1.jsx)(layout_1.Spacing, { y: 2, block: true })] }));
    }
    if (jobCount === 0) {
        return ((0, jsx_runtime_1.jsxs)("div", { style: explainer, children: [(0, jsx_runtime_1.jsx)(layout_1.Spacing, { y: 5 }), (0, jsx_runtime_1.jsx)("div", { style: errorExplanation, children: "No renders in the queue." }), (0, jsx_runtime_1.jsx)(layout_1.Spacing, { y: 2, block: true })] }));
    }
    return ((0, jsx_runtime_1.jsx)("div", { style: renderQueue, className: ['css-reset', is_menu_item_1.VERTICAL_SCROLLBAR_CLASSNAME].join(' '), children: jobs.map((j, index) => {
            return ((0, jsx_runtime_1.jsx)("div", { style: index === jobs.length - 1 ? undefined : separatorStyle, children: (0, jsx_runtime_1.jsx)(RenderQueueItem_1.RenderQueueItem, { job: j }) }, j.id));
        }) }));
};
exports.RenderQueue = RenderQueue;
