// @remotion/cli/dist/get-cli-options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCliOptions = exports.getAndValidateAbsoluteOutputFile = void 0;
const renderer_1 = require("@remotion/renderer");
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const config_1 = require("./config");
const get_env_1 = require("./get-env");
const get_input_props_1 = require("./get-input-props");
const log_1 = require("./log");
const getAndValidateFrameRange = () => {
    const frameRange = config_1.ConfigInternals.getRange();
    if (typeof frameRange === 'number') {
        log_1.Log.warn('Selected a single frame. Assuming you want to output an image.');
        log_1.Log.warn(`If you want to render a video, pass a range:  '--frames=${frameRange}-${frameRange}'.`);
        log_1.Log.warn("To dismiss this message, add the '--sequence' flag explicitly.");
    }
    return frameRange;
};
const getBrowser = () => { var _a; return (_a = config_1.ConfigInternals.getBrowser()) !== null && _a !== void 0 ? _a : renderer_1.RenderInternals.DEFAULT_BROWSER; };
const getAndValidateAbsoluteOutputFile = (relativeOutputLocation, overwrite) => {
    const absoluteOutputFile = node_path_1.default.resolve(process.cwd(), relativeOutputLocation);
    if (node_fs_1.default.existsSync(absoluteOutputFile) && !overwrite) {
        log_1.Log.error(`File at ${absoluteOutputFile} already exists. Use --overwrite to overwrite.`);
        process.exit(1);
    }
    return absoluteOutputFile;
};
exports.getAndValidateAbsoluteOutputFile = getAndValidateAbsoluteOutputFile;
const getAndValidateShouldOutputImageSequence = ({ frameRange, }) => {
    const shouldOutputImageSequence = config_1.ConfigInternals.getShouldOutputImageSequence(frameRange);
    return shouldOutputImageSequence;
};
const getCrf = (shouldOutputImageSequence) => {
    const crf = shouldOutputImageSequence
        ? null
        : config_1.ConfigInternals.getCrfOrUndefined();
    return crf;
};
const getProResProfile = () => {
    const proResProfile = config_1.ConfigInternals.getProResProfile();
    return proResProfile;
};
const getAndValidateBrowser = async (browserExecutable) => {
    const browser = getBrowser();
    try {
        await renderer_1.RenderInternals.ensureLocalBrowser(browser, browserExecutable);
    }
    catch (err) {
        log_1.Log.error('Could not download a browser for rendering frames.');
        log_1.Log.error(err);
        process.exit(1);
    }
    return browser;
};
const getCliOptions = async (options) => {
    var _a;
    const frameRange = getAndValidateFrameRange();
    const shouldOutputImageSequence = options.type === 'still'
        ? true
        : getAndValidateShouldOutputImageSequence({
            frameRange,
        });
    const overwrite = config_1.ConfigInternals.getShouldOverwrite({
        defaultValue: !options.isLambda,
    });
    const crf = getCrf(shouldOutputImageSequence);
    const videoBitrate = config_1.ConfigInternals.getVideoBitrate();
    const pixelFormat = config_1.ConfigInternals.getPixelFormat();
    const proResProfile = getProResProfile();
    const browserExecutable = config_1.ConfigInternals.getBrowserExecutable();
    const scale = config_1.ConfigInternals.getScale();
    const port = config_1.ConfigInternals.getServerPort();
    const chromiumOptions = {
        disableWebSecurity: config_1.ConfigInternals.getChromiumDisableWebSecurity(),
        ignoreCertificateErrors: config_1.ConfigInternals.getIgnoreCertificateErrors(),
        headless: config_1.ConfigInternals.getChromiumHeadlessMode(),
        gl: (_a = config_1.ConfigInternals.getChromiumOpenGlRenderer()) !== null && _a !== void 0 ? _a : renderer_1.RenderInternals.DEFAULT_OPENGL_RENDERER,
        userAgent: config_1.ConfigInternals.getChromiumUserAgent(),
    };
    const everyNthFrame = config_1.ConfigInternals.getEveryNthFrame();
    const numberOfGifLoops = config_1.ConfigInternals.getNumberOfGifLoops();
    const concurrency = config_1.ConfigInternals.getConcurrency();
    const height = config_1.ConfigInternals.getHeight();
    const width = config_1.ConfigInternals.getWidth();
    renderer_1.RenderInternals.validateConcurrency(concurrency, 'concurrency');
    return {
        puppeteerTimeout: config_1.ConfigInternals.getCurrentPuppeteerTimeout(),
        concurrency,
        frameRange,
        shouldOutputImageSequence,
        inputProps: (0, get_input_props_1.getInputProps)(null),
        envVariables: await (0, get_env_1.getEnvironmentVariables)(null),
        jpegQuality: config_1.ConfigInternals.getJpegQuality(),
        browser: await getAndValidateBrowser(browserExecutable),
        crf,
        pixelFormat,
        proResProfile,
        everyNthFrame,
        numberOfGifLoops,
        stillFrame: config_1.ConfigInternals.getStillFrame(),
        browserExecutable,
        logLevel: config_1.ConfigInternals.Logging.getLogLevel(),
        scale,
        chromiumOptions,
        overwrite,
        port: port !== null && port !== void 0 ? port : null,
        muted: config_1.ConfigInternals.getMuted(),
        enforceAudioTrack: config_1.ConfigInternals.getEnforceAudioTrack(),
        publicDir: config_1.ConfigInternals.getPublicDir(),
        ffmpegOverride: config_1.ConfigInternals.getFfmpegOverrideFunction(),
        audioBitrate: config_1.ConfigInternals.getAudioBitrate(),
        videoBitrate,
        height,
        width,
        configFileImageFormat: config_1.ConfigInternals.getUserPreferredVideoImageFormat(),
    };
};
exports.getCliOptions = getCliOptions;
