// @remotion/bundler/dist/homepage/homepage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Homepage = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const bundle_mode_1 = require("../bundle-mode");
const renderEntry_1 = require("../renderEntry");
const container = {
    width: 800,
    margin: 'auto',
    paddingLeft: 20,
    paddingRight: 20,
    fontFamily: 'sans-serif',
    lineHeight: 1.5,
};
const pre = {
    display: 'block',
    backgroundColor: '#f7f7f7',
    whiteSpace: 'nowrap',
    padding: 16,
    fontFamily: 'monospace',
    borderRadius: 5,
    fontSize: 15,
    overflowX: 'auto',
};
const AvailableCompositions = () => {
    const [state, setComps] = (0, react_1.useState)({
        type: 'not-initialized',
    });
    (0, react_1.useEffect)(() => {
        if ((0, bundle_mode_1.getBundleMode)().type !== 'evaluation') {
            return;
        }
        let timeout = null;
        const check = () => {
            if (window.remotion_renderReady === true) {
                setComps({ type: 'loading' });
                try {
                    const newComps = window.remotion_getCompositionNames();
                    setComps({ type: 'loaded', comps: newComps });
                }
                catch (err) {
                    setComps({ type: 'error', error: err });
                }
            }
            else {
                timeout = setTimeout(check, 250);
            }
        };
        check();
        return () => {
            if (!timeout) {
                return;
            }
            clearTimeout(timeout);
        };
    }, []);
    const showComps = (0, react_1.useCallback)(() => {
        (0, renderEntry_1.setBundleModeAndUpdate)({ type: 'evaluation' });
    }, []);
    if ((0, bundle_mode_1.getBundleMode)().type !== 'evaluation') {
        return ((0, jsx_runtime_1.jsx)("button", { type: "button", onClick: showComps, children: "Click here to see a list of available compositions." }));
    }
    if (state.type === 'loading') {
        return (0, jsx_runtime_1.jsx)("div", { children: state === null ? (0, jsx_runtime_1.jsx)("p", { children: "Loading compositions..." }) : null });
    }
    if (state.type === 'error') {
        return (0, jsx_runtime_1.jsxs)("div", { children: ["Error loading compositions: ", state.error.stack] });
    }
    if (state.type === 'not-initialized') {
        return (0, jsx_runtime_1.jsx)("div", { children: "Not initialized" });
    }
    return ((0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsx)("ul", { children: state === null
                ? []
                : state.comps.map((c) => {
                    return (0, jsx_runtime_1.jsx)("li", { children: c }, c);
                }) }) }));
};
const TestCORS = () => {
    const [serveUrl, setServeUrl] = (0, react_1.useState)('');
    const [result, setResult] = (0, react_1.useState)('');
    const handleServeUrl = (0, react_1.useCallback)((e) => {
        setServeUrl(e.target.value);
    }, []);
    const isCORSWorking = (0, react_1.useCallback)(async (e) => {
        e.preventDefault();
        try {
            const response = await fetch(serveUrl, { mode: 'cors' });
            if (response.ok) {
                setResult(`CORS is enabled on this URL: ${serveUrl}`);
            }
            else {
                setResult('URL does not support CORS - See DevTools console for more details');
            }
        }
        catch (error) {
            setResult('URL does not support CORS - See DevTools console for more details');
        }
    }, [serveUrl]);
    return ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsxs)("p", { children: ["Quickly test if a URL is supported being loaded on origin", ' ', (0, jsx_runtime_1.jsx)("code", { children: window.location.origin }), ". Enter the URL of an asset below."] }), result ? (0, jsx_runtime_1.jsx)("p", { className: "result", children: result }) : null, (0, jsx_runtime_1.jsxs)("form", { onSubmit: isCORSWorking, children: [(0, jsx_runtime_1.jsx)("label", { htmlFor: "serveurl", children: (0, jsx_runtime_1.jsx)("input", { placeholder: "Enter URL", type: "text", name: "serveurl", value: serveUrl, onChange: handleServeUrl }) }), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("button", { type: "submit", children: "Test CORS" })] })] }));
};
const Homepage = () => {
    const url = window.location.origin + window.location.pathname;
    return ((0, jsx_runtime_1.jsxs)("div", { style: container, children: [(0, jsx_runtime_1.jsx)("h1", { children: "Remotion Bundle" }), "This is a website which contains a bundled Remotion video. You can render videos based on this URL.", (0, jsx_runtime_1.jsx)("h2", { children: "Available compositions" }), (0, jsx_runtime_1.jsx)(AvailableCompositions, {}), (0, jsx_runtime_1.jsx)("h2", { children: "How to render" }), "Locally: ", (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsxs)("div", { style: pre, children: ["npx remotion render ", url, " ", '<comp-name> <output-location>'] }), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("br", {}), "With Remotion Lambda: ", (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsxs)("div", { style: pre, children: ["npx remotion lambda render ", url, " ", '<comp-name>'] }), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsxs)("p", { children: ["You can also render still images, and use the Node.JS APIs", ' ', (0, jsx_runtime_1.jsx)("code", { children: "getCompositions()" }), ", ", (0, jsx_runtime_1.jsx)("code", { children: "renderMedia()" }), ",", ' ', (0, jsx_runtime_1.jsx)("code", { children: "renderMediaOnLambda()" }), ", ", (0, jsx_runtime_1.jsx)("code", { children: "renderStill()" }), " and", ' ', (0, jsx_runtime_1.jsx)("code", { children: "renderStillOnLambda()" }), " with this URL."] }), (0, jsx_runtime_1.jsxs)("p", { children: ["Visit", ' ', (0, jsx_runtime_1.jsx)("a", { href: "https://remotion.dev/docs", target: "_blank", children: "remotion.dev/docs" }), ' ', "to read the documentation."] }), (0, jsx_runtime_1.jsx)("h2", { children: "CORS testing tool" }), (0, jsx_runtime_1.jsx)(TestCORS, {}), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("br", {}), (0, jsx_runtime_1.jsx)("br", {})] }));
};
exports.Homepage = Homepage;
