// @remotion/cli/dist/editor/components/SidebarRenderButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SidebarRenderButton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const client_1 = require("@remotion/renderer/client");
const react_1 = require("react");
const remotion_1 = require("remotion");
const get_default_out_name_1 = require("../../get-default-out-name");
const get_default_video_contexts_1 = require("../../preview-server/render-queue/get-default-video-contexts");
const client_id_1 = require("../helpers/client-id");
const render_1 = require("../icons/render");
const modals_1 = require("../state/modals");
const InlineAction_1 = require("./InlineAction");
const SidebarRenderButton = ({ composition, visible }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const iconStyle = (0, react_1.useMemo)(() => {
        return {
            style: {
                height: 12,
            },
        };
    }, []);
    const connectionStatus = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx).type;
    const { props } = (0, react_1.useContext)(remotion_1.Internals.EditorPropsContext);
    const isVideo = typeof composition.durationInFrames === 'undefined'
        ? true
        : composition.durationInFrames > 1;
    const onClick = (0, react_1.useCallback)((e) => {
        var _a;
        const defaults = window.remotion_renderDefaults;
        if (!defaults) {
            throw new Error('expected defaults');
        }
        e.stopPropagation();
        const { initialAudioCodec, initialRenderType, initialVideoCodec } = (0, get_default_video_contexts_1.getDefaultCodecs)({
            defaultCodec: defaults.codec,
            isStill: !isVideo,
        });
        setSelectedModal({
            type: 'render',
            compositionId: composition.id,
            initialFrame: 0,
            initialVideoImageFormat: defaults.videoImageFormat,
            initialStillImageFormat: defaults.stillImageFormat,
            initialJpegQuality: defaults.jpegQuality,
            initialScale: defaults.scale,
            initialVerbose: defaults.logLevel === 'verbose',
            initialOutName: (0, get_default_out_name_1.getDefaultOutLocation)({
                compositionName: composition.id,
                defaultExtension: isVideo
                    ? client_1.BrowserSafeApis.getFileExtensionFromCodec(initialVideoCodec, defaults.audioCodec)
                    : defaults.stillImageFormat,
                type: 'asset',
            }),
            initialVideoCodecForAudioTab: initialAudioCodec,
            initialRenderType,
            initialVideoCodecForVideoTab: initialVideoCodec,
            initialConcurrency: defaults.concurrency,
            maxConcurrency: defaults.maxConcurrency,
            minConcurrency: defaults.minConcurrency,
            initialMuted: defaults.muted,
            initialEnforceAudioTrack: defaults.enforceAudioTrack,
            initialProResProfile: defaults.proResProfile,
            initialPixelFormat: defaults.pixelFormat,
            initialAudioBitrate: defaults.audioBitrate,
            initialVideoBitrate: defaults.videoBitrate,
            initialEveryNthFrame: defaults.everyNthFrame,
            initialNumberOfGifLoops: defaults.numberOfGifLoops,
            initialDelayRenderTimeout: defaults.delayRenderTimeout,
            initialAudioCodec: defaults.audioCodec,
            initialEnvVariables: window.process.env,
            initialDisableWebSecurity: defaults.disableWebSecurity,
            initialOpenGlRenderer: defaults.openGlRenderer,
            initialHeadless: defaults.headless,
            initialIgnoreCertificateErrors: defaults.ignoreCertificateErrors,
            defaultProps: (_a = props[composition.id]) !== null && _a !== void 0 ? _a : composition.defaultProps,
            inFrameMark: null,
            outFrameMark: null,
        });
    }, [composition.defaultProps, composition.id, isVideo, props, setSelectedModal]);
    const renderAction = (0, react_1.useCallback)((color) => {
        return (0, jsx_runtime_1.jsx)(render_1.ThinRenderIcon, { fill: color, svgProps: iconStyle });
    }, [iconStyle]);
    if (!visible || connectionStatus !== 'connected') {
        return null;
    }
    return (0, jsx_runtime_1.jsx)(InlineAction_1.InlineAction, { renderAction: renderAction, onClick: onClick });
};
exports.SidebarRenderButton = SidebarRenderButton;
