// @remotion/cli/dist/editor/components/TopPanel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopPanel = exports.useResponsiveSidebarStatus = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const use_breakpoint_1 = require("../helpers/use-breakpoint");
const sidebar_1 = require("../state/sidebar");
const CanvasOrLoading_1 = require("./CanvasOrLoading");
const CompositionSelector_1 = require("./CompositionSelector");
const CurrentCompositionSideEffects_1 = require("./CurrentCompositionSideEffects");
const InitialCompositionLoader_1 = require("./InitialCompositionLoader");
const MenuToolbar_1 = require("./MenuToolbar");
const PreviewToolbar_1 = require("./PreviewToolbar");
const RightPanel_1 = require("./RightPanel");
const SplitterContainer_1 = require("./Splitter/SplitterContainer");
const SplitterElement_1 = require("./Splitter/SplitterElement");
const SplitterHandle_1 = require("./Splitter/SplitterHandle");
const container = {
    height: '100%',
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
};
const row = {
    display: 'flex',
    flexDirection: 'row',
    flex: 1,
};
const canvasContainer = {
    flex: 1,
    display: 'flex',
};
const leftContainer = {
    flex: 1,
    display: 'flex',
    maxWidth: '100%',
};
const useResponsiveSidebarStatus = () => {
    const { sidebarCollapsedStateLeft } = (0, react_1.useContext)(sidebar_1.SidebarContext);
    const responsiveLeftStatus = (0, use_breakpoint_1.useBreakpoint)(1200) ? 'collapsed' : 'expanded';
    const actualStateLeft = (0, react_1.useMemo)(() => {
        if (sidebarCollapsedStateLeft === 'collapsed') {
            return 'collapsed';
        }
        if (sidebarCollapsedStateLeft === 'expanded') {
            return 'expanded';
        }
        return responsiveLeftStatus;
    }, [sidebarCollapsedStateLeft, responsiveLeftStatus]);
    return actualStateLeft;
};
exports.useResponsiveSidebarStatus = useResponsiveSidebarStatus;
const TopPanel = () => {
    const { setSidebarCollapsedState, sidebarCollapsedStateRight } = (0, react_1.useContext)(sidebar_1.SidebarContext);
    const actualStateLeft = (0, exports.useResponsiveSidebarStatus)();
    const actualStateRight = (0, react_1.useMemo)(() => {
        if (sidebarCollapsedStateRight === 'collapsed') {
            return 'collapsed';
        }
        return 'expanded';
    }, [sidebarCollapsedStateRight]);
    const onCollapseLeft = (0, react_1.useCallback)(() => {
        setSidebarCollapsedState({ left: 'collapsed', right: null });
    }, [setSidebarCollapsedState]);
    const onCollapseRight = (0, react_1.useCallback)(() => {
        setSidebarCollapsedState({ left: null, right: 'collapsed' });
    }, [setSidebarCollapsedState]);
    return ((0, jsx_runtime_1.jsxs)("div", { style: container, children: [(0, jsx_runtime_1.jsx)(InitialCompositionLoader_1.InitialCompositionLoader, {}), (0, jsx_runtime_1.jsx)(MenuToolbar_1.MenuToolbar, {}), (0, jsx_runtime_1.jsx)("div", { style: row, children: (0, jsx_runtime_1.jsxs)(SplitterContainer_1.SplitterContainer, { minFlex: 0.15, maxFlex: 0.4, defaultFlex: 0.2, id: "sidebar-to-preview", orientation: "vertical", children: [actualStateLeft === 'expanded' ? ((0, jsx_runtime_1.jsx)(SplitterElement_1.SplitterElement, { type: "flexer", children: (0, jsx_runtime_1.jsx)("div", { style: leftContainer, className: "css-reset", children: (0, jsx_runtime_1.jsx)(CompositionSelector_1.CompositionSelector, {}) }) })) : null, actualStateLeft === 'expanded' ? ((0, jsx_runtime_1.jsx)(SplitterHandle_1.SplitterHandle, { allowToCollapse: "left", onCollapse: onCollapseLeft })) : null, (0, jsx_runtime_1.jsx)(SplitterElement_1.SplitterElement, { type: "anti-flexer", children: (0, jsx_runtime_1.jsxs)(SplitterContainer_1.SplitterContainer, { minFlex: 0.5, maxFlex: 0.8, defaultFlex: 0.7, id: "canvas-to-right-sidebar", orientation: "vertical", children: [(0, jsx_runtime_1.jsx)(SplitterElement_1.SplitterElement, { type: "flexer", children: (0, jsx_runtime_1.jsx)("div", { style: canvasContainer, children: (0, jsx_runtime_1.jsx)(CanvasOrLoading_1.CanvasOrLoading, {}) }) }), actualStateRight === 'expanded' ? ((0, jsx_runtime_1.jsx)(SplitterHandle_1.SplitterHandle, { allowToCollapse: "right", onCollapse: onCollapseRight })) : null, actualStateRight === 'expanded' ? ((0, jsx_runtime_1.jsx)(SplitterElement_1.SplitterElement, { type: "anti-flexer", children: (0, jsx_runtime_1.jsx)(RightPanel_1.RightPanel, {}) })) : null] }) })] }) }), (0, jsx_runtime_1.jsx)(PreviewToolbar_1.PreviewToolbar, {}), (0, jsx_runtime_1.jsx)(CurrentCompositionSideEffects_1.CurrentCompositionKeybindings, {}), (0, jsx_runtime_1.jsx)(CurrentCompositionSideEffects_1.TitleUpdater, {})] }));
};
exports.TopPanel = TopPanel;
